File: packages/javascript/lib/string/getGroupKeys.ts
export function getGroupKeys(
	str: string,
	groupStart: string,
	groupEnd: string,
	ignoreValidation = false
): string[] {
	if (!str) return [];
	if (!groupStart || !groupEnd) throw new Error('`groupStart` and `groupEnd` must be non-empty.');
	if (groupStart === groupEnd) throw new Error('`groupStart` and `groupEnd` must be different.');
	if (groupStart.length === 1 && groupEnd.length !== 1) {
		throw new Error(
			'When `groupStart` is a single character, `groupEnd` must also be a single character.'
		);
	}
	if (groupEnd.length === 1 && groupStart.length !== 1) {
		throw new Error(
			'When `groupEnd` is a single character, `groupStart` must also be a single character.'
		);
	}

	const out: string[] = [];
	const n = str.length;
	const sLen = groupStart.length;
	const eLen = groupEnd.length;

	const isEscapedAt = (idx: number) => {
		let bs = 0;
		for (let k = idx - 1; k >= 0 && str[k] === '\\'; k--) bs++;
		return bs % 2 === 1;
	};

	const isSingle = sLen === 1 && eLen === 1;
	const sCh = isSingle ? groupStart : '';
	const eCh = isSingle ? groupEnd : '';

	const isValidStartAt = (i: number) => {
		if (i < 0 || i + sLen > n) return false;
		if (str.slice(i, i + sLen) !== groupStart) return false;
		if (isEscapedAt(i)) return false;
		if (isSingle) {
			if (i > 0 && str[i - 1] === sCh) return false;
			if (i + 1 < n && str[i + 1] === sCh) return false;
		}
		return true;
	};

	const isValidEndAt = (j: number) => {
		if (j < 0 || j + eLen > n) return false;
		if (str.slice(j, j + eLen) !== groupEnd) return false;
		if (isEscapedAt(j)) return false;
		if (isSingle) {
			if (j > 0 && str[j - 1] === eCh) return false;
			if (j + 1 < n && str[j + 1] === eCh) return false;
		}
		return true;
	};

	const isValidKey = (key: string) => {
		if (key.includes('\n') || key.includes('\r')) return false;
		if (key.includes(groupStart) || key.includes(groupEnd)) return false;
		if (ignoreValidation) return true;
		return /^[A-Za-z0-9_$-]*$/.test(key);
	};

	for (let i = 0; i <= n - sLen; i++) {
		if (!isValidStartAt(i)) continue;

		const begin = i + sLen;
		let endAt = -1;

		for (let j = begin; j <= n - eLen; j++) {
			if (isValidEndAt(j)) {
				endAt = j;
				break;
			}
		}

		if (endAt === -1) continue;

		const key = str.slice(begin, endAt);
		if (isValidKey(key)) out.push(key);

		i = endAt + eLen - 1;
	}

	return out;
}

File: packages/python/qsu/string/getGroupKeys.py
import re


def getGroupKeys(
	str: str, groupStart: str, groupEnd: str, ignoreValidation: bool = False
) -> list:
	if not str:
		return []
	if not groupStart or not groupEnd:
		raise ValueError('`groupStart` and `groupEnd` must be non-empty.')
	if groupStart == groupEnd:
		raise ValueError('`groupStart` and `groupEnd` must be different.')
	if len(groupStart) == 1 and len(groupEnd) != 1:
		raise ValueError(
			'When `groupStart` is a single character, `groupEnd` must also be a single character.'
		)
	if len(groupEnd) == 1 and len(groupStart) != 1:
		raise ValueError(
			'When `groupEnd` is a single character, `groupStart` must also be a single character.'
		)

	out = []
	n = len(str)
	sLen = len(groupStart)
	eLen = len(groupEnd)

	def isEscapedAt(idx):
		bs = 0
		k = idx - 1
		while k >= 0 and str[k] == '\\':
			bs += 1
			k -= 1
		return bs % 2 == 1

	isSingle = sLen == 1 and eLen == 1
	sCh = groupStart if isSingle else ''
	eCh = groupEnd if isSingle else ''

	def isValidStartAt(i):
		if i < 0 or i + sLen > n:
			return False
		if str[i : i + sLen] != groupStart:
			return False
		if isEscapedAt(i):
			return False
		if isSingle:
			if i > 0 and str[i - 1] == sCh:
				return False
			if i + 1 < n and str[i + 1] == sCh:
				return False
		return True

	def isValidEndAt(j):
		if j < 0 or j + eLen > n:
			return False
		if str[j : j + eLen] != groupEnd:
			return False
		if isEscapedAt(j):
			return False
		if isSingle:
			if j > 0 and str[j - 1] == eCh:
				return False
			if j + 1 < n and str[j + 1] == eCh:
				return False
		return True

	def isValidKey(key):
		if '\n' in key or '\r' in key:
			return False
		if groupStart in key or groupEnd in key:
			return False
		if ignoreValidation:
			return True
		return re.fullmatch(r'[A-Za-z0-9_$-]*', key) is not None

	i = 0
	while i <= n - sLen:
		if not isValidStartAt(i):
			i += 1
			continue

		begin = i + sLen
		endAt = -1

		j = begin
		while j <= n - eLen:
			if isValidEndAt(j):
				endAt = j
				break
			j += 1

		if endAt == -1:
			i += 1
			continue

		key = str[begin:endAt]
		if isValidKey(key):
			out.append(key)

		i = endAt + eLen - 1
		i += 1

	return out
